drivers/i3c/i3c_mcux.c

```diff
@@ -0,0 +1,2063 @@
+	base->MINTCLR = I3C_MINTCLR_SLVSTART_MASK;
+
+	data->ibi.addr[idx] = 0U;
+	data->ibi.num_addr -= 1U;
+
+	/* Tell target to disable IBI */
+	i3c_events.events = I3C_CCC_EVT_INTR;
+	ret = i3c_ccc_do_events_set(target, false, &i3c_events);
+	if (ret != 0) {
+		LOG_ERR("Error sending IBI DISEC for 0x%02x (%d)",
+			target->dynamic_addr, ret);
+
+		goto out;
+	}
+
+	mcux_i3c_ibi_rules_setup(data, base);
+
+out:
+	if (data->ibi.num_addr > 0U) {
+		/*
+		 * Enable controller to raise interrupt when a target
+		 * initiates IBI.
+		 */
+		base->MINTSET = I3C_MINTSET_SLVSTART_MASK;
+	}
+
+	return ret;
+}
+#endif /* CONFIG_I3C_USE_IBI */
+
+/**
+ * @brief Interrupt Service Routine
+ *
+ * Currently only services interrupts when any target initiates IBIs.
+ *
+ * @param dev Pointer to controller device driver instance.
+ */
+static void mcux_i3c_isr(const struct device *dev)
+{
+#ifdef CONFIG_I3C_USE_IBI
+	const struct mcux_i3c_config *config = dev->config;
+	I3C_Type *base = config->base;
+
+	/* Target initiated IBIs */
+	if (mcux_i3c_status_is_set(base, I3C_MSTATUS_SLVSTART_MASK)) {
+		/*
+		 * Disable further target initiated IBI interrupt
+		 * while we try to service the current one.
+		 */
+		base->MINTCLR = I3C_MINTCLR_SLVSTART_MASK;
+
+		/*
+		 * Handle IBI in workqueue.
+		 */
+		i3c_ibi_work_enqueue_cb(dev, mcux_i3c_ibi_work);
+	}
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+/**
+ * @brief Initialize I3C bus.
+ *
+ * @param dev Pointer to controller device driver instance.
+ *
+ * @return 0 if successful, negative if errors.
+ */
+static int mcux_i3c_bus_init(const struct device *dev)
+{
+	struct mcux_i3c_data *data = dev->data;
+
+	return i3c_bus_init(dev, &data->device_list);
+}
+
+/**
+ * @brief Initialize the hardware.
+ *
+ * @param dev Pointer to controller device driver instance.
+ */
+static int mcux_i3c_init(const struct device *dev)
+{
+	const struct mcux_i3c_config *config = dev->config;
+	struct mcux_i3c_data *data = dev->data;
+	I3C_Type *base = config->base;
+	i3c_master_config_t ctrl_config;
+	uint32_t clock_freq;
+	int error;
+	int ret = 0;
+
+	i3c_dev_list_init(&data->device_list);
+
+	CLOCK_SetClkDiv(kCLOCK_DivI3cClk, config->clk_div_pp);
+	CLOCK_SetClkDiv(kCLOCK_DivI3cSlowClk, config->clk_div_od);
+	CLOCK_SetClkDiv(kCLOCK_DivI3cTcClk, config->clk_div_tc);
+
+#ifdef CONFIG_PINCTRL
+	error = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (error) {
+		ret = error;
+		goto err_out;
+	}
+#endif
+
+	k_sem_init(&data->lock, 1, 1);
+	k_sem_init(&data->ibi_lock, 1, 1);
+
+	/* Get the clock frequency */
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys,
+				   &clock_freq)) {
+		ret = -EINVAL;
+		goto err_out;
+	}
+
+	/*
+	 * Default controller configuration to act as the primary
+	 * and active controller.
+	 */
+	I3C_MasterGetDefaultConfig(&ctrl_config);
+
+	/* Set SCL clock rate (in Hz) */
+	if (config->i2c_scl_hz != 0U) {
+		ctrl_config.baudRate_Hz.i2cBaud = config->i2c_scl_hz;
```

Comment:
So are speeds just fixed here? Once they're set in the DTS, it looks like there's no way to change them.

Reply:
Oops... I was so focused on getting IBI to work I forgot to add a `configure` function. Will add.

Reply:
Added a way to configure the controller.